### DISCUSSION TITLE
JSON Encoding Performance

### DISCUSSION THREAD

**Discussion by pschichtel** [osiegmar/logback-gelf#83] [Ideas]
As part of the performance work I've been doing on the project I work on, I ended up completely rewriting the GelfEncoder to avoid allocations and remove overhead.

https://gist.github.com/pschichtel/cb5556b9d5b681e1e77afabb749186e9

I implemented it using Jackson and Kotlin, since we are using those anyway, so I don't expect any of this to be merged.

This implementation is significantly faster and generates a lot fewer objects then the original one.

However while implementing this I noticed two "issues":

1. the gelf version constant is not accessible
2. the GelfUdpAppender (I assume the same to be the case for the others) expect a subclass of GelfEncoder

The former could easily be solved by making it a public constant, not so sure about the latter, but it is also not really a problem.

**Comment by pschichtel** [comment]
@osiegmar should I create issues for the two points I had? Would it be useful to get my encoder upstream, e.g. as a separate module? I could rewrite it in Java if there is interest in this.

**Comment by osiegmar** [comment]
@pschichtel Sorry for replying late. Currently I don't see a need for changing / enhancing this library's code base in order to tune its performance. I'm wondering what exact problems you had that led you to re-implement it?

**Reply by pschichtel** [reply]
@osiegmar the suggestion I made are not about performance, they are about customization. The first point is about the version constant, that's not accessibly (which I would have used), the second point is about the appender expecting a subclass of GelfEncoder, which seemed unnecessary at the time. Neither point is really critical.

This was the outcome of a profiling session which showed significant garbage in the encoder.

**Reply by osiegmar** [reply]
I see. Yes, extensibility was not considered while implementing those classes. I'll consider this for the next update.

Update: I created #89 for this – obviously this is not automatically linked in to this discussion.
